ext/liquid_c/vm.c

```diff
@@ -290,6 +290,18 @@ static VALUE vm_render_until_error(VALUE uncast_args)
                 break;
             }
 
+            // Filters instructions
+
+            case OP_APPEND:
+            {
+                VALUE *args_ptr = vm_stack_pop_n_use_in_place(vm, 2);
+                VALUE result = rb_str_plus(
+                    rb_obj_as_string(args_ptr[0]),
```

Comment:
It will also avoid going through the filter invocation logic in Ruby. I think it is the source of many allocations, and might be causing excessive GC on templates w/ lots of filters in loops. Here are top allocation spots from a slow shop (using https://github.com/ko1/allocation_tracer#allocation-tracing):

```
[["[\"/Users/ma/.gem/ruby/2.7.1/bundler/gems/liquid-077bf2a409b9/lib/liquid/variable.rb\", 86, :T_ARRAY]",
  [10560, 0, 3415, 0, 1, 0]],
 ["[\"/Users/ma/.gem/ruby/2.7.1/bundler/gems/liquid-077bf2a409b9/lib/liquid/variable.rb\", 84, :T_IMEMO]",
  [7016, 0, 2296, 0, 1, 0]],
 ["[\"/Users/ma/.gem/ruby/2.7.1/bundler/gems/liquid-077bf2a409b9/lib/liquid/standardfilters.rb\", 281, :T_STRING]",
  [5642, 18, 1871, 0, 1, 0]],
 ["[\"/Users/ma/.gem/ruby/2.7.1/bundler/gems/liquid-077bf2a409b9/lib/liquid/strainer_template.rb\", 43, :T_ARRAY]",
  [5280, 0, 1707, 0, 1, 0]],
 ["[\"/Users/ma/.gem/ruby/2.7.1/bundler/gems/liquid-077bf2a409b9/lib/liquid/variable.rb\", 132, :T_ARRAY]",
  [5280, 0, 1708, 0, 1, 0]],
```

All those are from filter invocation as far as I can tell.

RE `.to_s`. Aye! I was totally wrong with the duplication indeed, not sure where I got that. Thanks for clarifying.

But that gave me an idea. What if we extracted the string conversion into a `OP_TO_STRING` instruction (`stack_push(stack_pop.to_s)`). In the assembler, we always generate it before instructions that accept strings (also introduce a new `PUSH_STRING`, that does same thing as `PUSH_CONST` but acts as a type annotation):

```
{% a | append: 'x' | append: 'y' %}

FIND_VAR       a
TO_STRING
PUSH_STRING   'x'
TO_STRING
APPEND
PUSH_STRING   'y'
TO_STRING
APPEND
```

Then, we can optimize by removing all the subsequent `TO_STRING` since we know APPEND returns a string:

```
FIND_VAR       a
TO_STRING
PUSH_STRING   'x'
APPEND
PUSH_STRING   'y'
APPEND
```

That would remove a lot of the type checks in the VM loop.

Reply:
> It will also avoid going through the filter invocation logic in Ruby.

I think that is already the case for variable tags.  Once the `echo` and `assign` tags are compiled, then variable will be completely handled from the liquid VM, which will avoid those allocations that the ruby liquid gem does.

Reply:
We could definitely do some type inference to avoid type casting overhead in the common path in the future.  However, currently we don't do that statically at all, so I think there would be a lot of work before we could get rid of a lot of the type check/conversion overhead in the VM loop.

Reply:
> Once the `echo` and `assign` tags are compiled, then variable will be completely handled from the liquid VM

Ah! I missed that the expression part of the assign was not compiled to bytecode yet.

I think once assign is implemented we'll see a huge different in allocations and speed for some slow shops. So I'll close this PR. I more general method lookup caching approach would probably be better.